File: frontend/src/types/research.ts
// ── Enums ─────────────────────────────────────────────────────────────────────

export type AgentStatus = 'pending' | 'running' | 'done' | 'error' | 'skipped'
export type ResearchStatus =
  | 'queued'
  | 'planning'
  | 'researching'
  | 'critiquing'
  | 'refining'
  | 'synthesizing'
  | 'complete'
  | 'failed'
export type SourceType = 'web' | 'arxiv' | 'semantic_scholar' | 'pdf' | 'internal'

export type SSEEventType =
  | 'pipeline_start'
  | 'agent_start'
  | 'agent_progress'
  | 'agent_complete'
  | 'agent_error'
  | 'sources_found'
  | 'refinement_loop'
  | 'report_section'
  | 'pipeline_complete'
  | 'pipeline_error'
  | 'heartbeat'

// ── Domain models ─────────────────────────────────────────────────────────────

export interface Source {
  id: string
  url: string
  title: string
  snippet: string
  source_type: SourceType
  published_date: string | null
  credibility_score: number
  retrieved_at: string
  authors: string[]
  citation_count: number
}

export interface ReportSection {
  heading: string
  content: string
  confidence: number
  citations: Citation[]
  word_count: number
  gaps_noted: string[]
}

export interface Citation {
  source_id: string
  claim: string
  quote: string
  page_number: number | null
}

export interface ResearchReport {
  id: string
  title: string
  executive_summary: string
  sections: ReportSection[]
  all_sources: Source[]
  overall_confidence: number
  total_sources: number
  word_count: number
  generated_at: string
  refinement_iterations: number
  model_versions: Record<string, string>
}

export interface AgentTrace {
  agent_id: string
  status: AgentStatus
  started_at: string | null
  completed_at: string | null
  duration_ms: number | null
  token_usage: Record<string, number>
  tool_calls: Record<string, unknown>[]
  error: string | null
}

export interface SubTask {
  id: string
  heading: string
  objective: string
  assigned_to: string
  allowed_sources: SourceType[]
  scope_rules: string[]
  priority: number
  status: AgentStatus
  findings: string[]
  sources: Source[]
}

export interface PipelineState {
  session_id: string
  user_id: string | null
  query: string
  status: ResearchStatus
  sub_tasks: SubTask[]
  agent_traces: Record<string, AgentTrace>
  refinement_count: number
  report: ResearchReport | null
  created_at: string
  updated_at: string
  error_message: string | null
  metadata: Record<string, unknown>
}

// ── SSE event envelope ────────────────────────────────────────────────────────

export interface SSEEvent<T = Record<string, unknown>> {
  event: SSEEventType
  session_id: string
  timestamp: string
  data: T & { agent?: string }
}

// ── Specific event payloads ───────────────────────────────────────────────────

export interface AgentStartData {
  agent_role: string
  message: string
}

export interface AgentProgressData {
  message: string
  agent_role?: string
  analysis?: Record<string, unknown>
  subtasks?: string[]
  gaps?: string[]
  refinement_needed?: boolean
  sources?: number
  confidence?: number
}

export interface SourcesFoundData {
  total_sources: number
  by_type: Partial<Record<SourceType, number>>
}

export interface RefinementLoopData {
  iteration: number
  max_iterations: number
  targets: string[]
}

export interface PipelineCompleteData {
  session_id: string
  duration_seconds: number
  tokens_used: number
  report_sections: number
  total_sources: number
  overall_confidence: number
  report?: ResearchReport
}

// ── API types ─────────────────────────────────────────────────────────────────

export interface ResearchRequest {
  query: string
  depth: 'quick' | 'standard' | 'deep'
  max_sources: number
  include_arxiv: boolean
  language: string
}

export interface ResearchResponse {
  session_id: string
  status: ResearchStatus
  message: string
}

// ── UI state ──────────────────────────────────────────────────────────────────

export interface AgentUIState {
  role: string
  displayName: string
  status: AgentStatus
  message: string
  durationMs: number | null
  tokensUsed: number | null
  findings: string[]
  sourcesCount: number
}

export interface PipelineUIState {
  sessionId: string | null
  query: string
  status: ResearchStatus | 'idle'
  agents: Record<string, AgentUIState>
  report: ResearchReport | null
  totalSources: number
  refinementCount: number
  durationSeconds: number | null
  tokensUsed: number | null
  error: string | null
  events: SSEEvent[]
}

File: backend/app/schemas/research.py
"""
Pydantic v2 schemas — the contract between frontend, API, and agents.
All pipeline state is typed end-to-end.
"""
from __future__ import annotations
from datetime import datetime
from enum import Enum
from typing import Any
from uuid import UUID, uuid4
from pydantic import BaseModel, Field, HttpUrl, field_validator


# ── Enums ────────────────────────────────────────────────────────────────────

class AgentStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    DONE = "done"
    ERROR = "error"
    SKIPPED = "skipped"


class ResearchStatus(str, Enum):
    QUEUED = "queued"
    PLANNING = "planning"
    RESEARCHING = "researching"
    CRITIQUING = "critiquing"
    REFINING = "refining"
    SYNTHESIZING = "synthesizing"
    COMPLETE = "complete"
    FAILED = "failed"


class SourceType(str, Enum):
    WEB = "web"
    ARXIV = "arxiv"
    SEMANTIC_SCHOLAR = "semantic_scholar"
    PDF = "pdf"
    INTERNAL = "internal"


# ── Source & Citation ─────────────────────────────────────────────────────────

class Source(BaseModel):
    id: UUID = Field(default_factory=uuid4)
    url: str
    title: str
    snippet: str = ""
    source_type: SourceType = SourceType.WEB
    published_date: datetime | None = None
    credibility_score: float = Field(default=0.5, ge=0.0, le=1.0)
    retrieved_at: datetime = Field(default_factory=datetime.utcnow)
    authors: list[str] = []
    citation_count: int = 0

    @field_validator("credibility_score")
    @classmethod
    def round_score(cls, v: float) -> float:
        return round(v, 3)


class Citation(BaseModel):
    source_id: UUID
    claim: str
    quote: str = ""
    page_number: int | None = None


# ── Sub-task ──────────────────────────────────────────────────────────────────

class SubTask(BaseModel):
    id: UUID = Field(default_factory=uuid4)
    heading: str
    objective: str
    assigned_to: str  # "researcher_a" | "researcher_b"
    allowed_sources: list[SourceType] = []
    scope_rules: list[str] = []
    priority: int = Field(default=1, ge=1, le=5)
    status: AgentStatus = AgentStatus.PENDING
    findings: list[str] = []
    sources: list[Source] = []


# ── Report sections ───────────────────────────────────────────────────────────

class ReportSection(BaseModel):
    heading: str
    content: str
    confidence: float = Field(ge=0.0, le=1.0)
    citations: list[Citation] = []
    word_count: int = 0
    gaps_noted: list[str] = []


class ResearchReport(BaseModel):
    id: UUID = Field(default_factory=uuid4)
    title: str
    executive_summary: str
    sections: list[ReportSection]
    all_sources: list[Source]
    overall_confidence: float = Field(ge=0.0, le=1.0)
    total_sources: int = 0
    word_count: int = 0
    generated_at: datetime = Field(default_factory=datetime.utcnow)
    refinement_iterations: int = 0
    model_versions: dict[str, str] = {}


# ── Pipeline state (persisted in Redis + DB) ──────────────────────────────────

class AgentTrace(BaseModel):
    agent_id: str
    status: AgentStatus
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = None
    token_usage: dict[str, int] = {}
    tool_calls: list[dict[str, Any]] = []
    error: str | None = None


class PipelineState(BaseModel):
    session_id: UUID = Field(default_factory=uuid4)
    user_id: UUID | None = None
    query: str
    status: ResearchStatus = ResearchStatus.QUEUED
    sub_tasks: list[SubTask] = []
    agent_traces: dict[str, AgentTrace] = {}
    refinement_count: int = 0
    report: ResearchReport | None = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    error_message: str | None = None
    metadata: dict[str, Any] = {}


# ── SSE event envelope ────────────────────────────────────────────────────────

class SSEEventType(str, Enum):
    PIPELINE_START = "pipeline_start"
    AGENT_START = "agent_start"
    AGENT_PROGRESS = "agent_progress"
    AGENT_COMPLETE = "agent_complete"
    AGENT_ERROR = "agent_error"
    SOURCES_FOUND = "sources_found"
    REFINEMENT_LOOP = "refinement_loop"
    REPORT_SECTION = "report_section"
    PIPELINE_COMPLETE = "pipeline_complete"
    PIPELINE_ERROR = "pipeline_error"
    HEARTBEAT = "heartbeat"


class SSEEvent(BaseModel):
    event: SSEEventType
    session_id: str
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    data: dict[str, Any] = {}

    def to_sse_string(self) -> str:
        import json
        payload = self.model_dump_json()
        return f"event: {self.event.value}\ndata: {payload}\n\n"


# ── API request/response ──────────────────────────────────────────────────────

class ResearchRequest(BaseModel):
    query: str = Field(..., min_length=10, max_length=2000)
    depth: Literal["quick", "standard", "deep"] = "standard"  # type: ignore[name-defined]
    max_sources: int = Field(default=20, ge=5, le=50)
    include_arxiv: bool = True
    language: str = "en"


class ResearchResponse(BaseModel):
    session_id: str
    status: ResearchStatus
    message: str = ""


class HealthResponse(BaseModel):
    status: str
    version: str
    environment: str
    checks: dict[str, bool] = {}


# fix forward ref
from typing import Literal  # noqa: E402 — after enum definitions
